tests/openedx_tagging/core/tagging/test_views.py

```diff
@@ -0,0 +1,299 @@
+
+        create_data = {
+            "name": "taxonomy_data_3",
+            "description": "This is a description",
+            "enabled": False,
+            "required": True,
+            "allow_multiple": True,
+            "taxonomy_class": 'openedx_tagging.core.tagging.models.Taxonomy'
+        }
+
+        self.client.force_authenticate(user=self.staff)
+        response = self.client.post(url, create_data)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        self.assertGreaterEqual(response.data.items(), create_data.items())
+        url = reverse(
+            "oel_tagging:taxonomy-detail", kwargs={"id": response.data["id"]}
+        )
+        response = self.client.get(url)
+        self.assertGreaterEqual(response.data.items(), create_data.items())
+
+    @ddt.data(
+        {},
+        {"name": "Error taxonomy 2", "required": "Invalid value"},
+        {"name": "Error taxonomy 3", "enabled": "Invalid value"},
+    )
+    def test_create_taxonomy_error(self, create_data):
+        url = reverse("oel_tagging:taxonomy-list")
+
+        self.client.force_authenticate(user=self.staff)
+        response = self.client.post(url, create_data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    @ddt.data({"name": "System defined taxonomy", "system_defined": True})
+    def test_create_taxonomy_system_defined(self, create_data):
+        """
+        Cannont create a taxonomy with system_defined=true
+        """
+        url = reverse("oel_tagging:taxonomy-list")
+
+        self.client.force_authenticate(user=self.staff)
+        response = self.client.post(url, create_data)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(response.data["system_defined"], False)
+
+    @ddt.data(
+        (None, status.HTTP_403_FORBIDDEN),
+        ("user", status.HTTP_403_FORBIDDEN),
+        ("staff", status.HTTP_200_OK),
+    )
+    @ddt.unpack
+    def test_update_taxonomy(self, user_attr, expected_status):
+        taxonomy = Taxonomy.objects.create(
+            name="test update taxonomy",
+            description="taxonomy description",
+            enabled=False,
+        )
+        taxonomy.save()
+
+        url = reverse("oel_tagging:taxonomy-detail", kwargs={"id": taxonomy.id})
+        user = getattr(self, user_attr) if user_attr else None
+        if user:
+            self.client.force_authenticate(user=user)
+        response = self.client.put(url, {"name": "new name"})
+        self.assertEqual(response.status_code, expected_status)
+
+        # If we were able to update the taxonomy, check if the name changed
+        if status.is_success(expected_status):
+            response = self.client.get(url)
+            self.assertEqual(response.data["name"], "new name")
+            self.assertEqual(response.data["enabled"], False)
+            self.assertEqual(response.data["description"], "taxonomy description")
```

Comment:
Yes, it's worthwhile checking that `patch` only updates what we think it does.
But to do this properly, you need to check the value of fields which you _haven't updated with the PATCH_ as well.
The `check_taxonomy` method I proposed uses the default values expected for a new taxonomy in its method kwargs, which you can override by passing in your `**update_args` or whatever you sent to the `put`/`patch`.

A simplified example:

```python
def check_taxonomy(data, id, description="", allow_free_text=False, enabled=True):
   assert data["id"] == id
   assert data["description"] == description
   assert data["enabled"] == enabled
   assert data["allow_free_text"] == allow_free_text

update_args = { "allow_free_text" = True, description = "new description"}
check_taxonomy(data, taxonomy.id, **update_args)
# ^ this will compare:
* assert data["description"] ==  "new description"
* assert data["enabled"] == True  # falls back to check_taxonomy param's default, because it's not in update_args
* assert data["allow_free_text"] == True
```

Reply:
Sure!